proto/p4/p4types.proto

```diff
@@ -0,0 +1,207 @@
+ * | enum               | error     | error        | allowed         |
+ * | header             | error     | allowed      | allowed         |
+ * | header stack       | error     | error        | allowed         |
+ * | header_union       | error     | error        | allowed         |
+ * | struct             | error     | error        | allowed         |
+ * | tuple              | error     | error        | allowed         |
+ * |--------------------|-----------|--------------|-----------------|
+ */
+
+// These P4 types (struct, header_type, header_union and enum) are guaranteed to
+// be "named" (e.g. you cannot use an anonymous struct to declare a variable,
+// like in C). Instead of duplicating the type spec for these every time the
+// type is used, we include the type spec once in this P4TypeInfo message and
+// refer to the types by name in the P4DataTypeSpec message. We also support
+// annotations for these type specs which can be useful, e.g. to identify
+// well-known headers (such as ipv4).
+message P4TypeInfo {
+  map<string, P4StructTypeSpec> structs = 1;
+  map<string, P4HeaderTypeSpec> headers = 2;
+  map<string, P4HeaderUnionTypeSpec> header_unions = 3;
+  map<string, P4EnumTypeSpec> enums = 4;
+  P4ErrorTypeSpec error = 5;
+}
+
+// Describes a P4_16 type.
+message P4DataTypeSpec {
+  oneof type_spec {
+    P4BitstringLikeTypeSpec bitstring = 1;
+    P4BoolType bool = 2;
+    // From the P4_16 spec: "A tuple is similar to a struct, in that it holds
+    // multiple values. Unlike a struct type, tuples have no named fields."
+    P4TupleTypeSpec tuple = 3;
+    P4NamedType struct = 4;
+    P4NamedType header = 5;
+    P4NamedType header_union = 6;
+    P4HeaderStackTypeSpec header_stack = 7;
+    P4HeaderUnionStackTypeSpec header_union_stack = 8;
+    P4NamedType enum = 9;
+    P4ErrorType error = 10;
+  }
+}
+
+message P4NamedType {
+  string name = 1;
+}
+
+// Empty message as no type information needed, just used as a placeholder in
+// the oneof to identify boolean types.
+message P4BoolType { }
+
+message P4ErrorType { }
+
+message P4BitstringLikeTypeSpec {
+  oneof type_spec {
+    P4BitTypeSpec bit = 1;  // bit<W>
+    P4IntTypeSpec int = 2;  // int<W>
+    P4VarbitTypeSpec varbit = 3;  // varbit<W>
+  }
+  // Useful to identify well-know types, such as IP address or Ethernet MAC
+  // address.
+  repeated string annotations = 4;
+}
+
+message P4BitTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4IntTypeSpec {
+  int32 bitwidth = 1;
+}
+
+message P4VarbitTypeSpec {
+  int32 max_bitwidth = 1;
+}
+
+message P4TupleTypeSpec {
+  repeated P4DataTypeSpec members = 1;
+}
+
+message P4StructTypeSpec {
+  message Member {
+    string name = 1;
+    P4DataTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
+  repeated string annotations = 2;
+}
+
+message P4HeaderTypeSpec {
+  message Member {
+    string name = 1;
+    P4BitstringLikeTypeSpec type_spec = 2;
+  }
+  repeated Member members = 1;
```

Comment:
What needs to be said about the order of members?  Can an implementation expect that for any valid value of N, the offset of members(N) in the header is greater than the offset of members(N-1) from the start of the header?

Reply:
I just updated the documentation to specify that the order of the repeated "members" field for compound types must match the order in the corresponding P4 declaration. I didn't add comments to the proto, since I would need to have the same comment in a dozen place.